_posts/2021-07-22-randomize-slack-2.md

```diff
@@ -0,0 +1,117 @@
+---
+layout:      posts
+background:  shorterBackground
+title:       "Randomize your Slack Avatar: Part 2"
+subtitle:    ""
+date:        2021-07-22 12:00
+author:      "Drew Garrett"
+headerImg:   "/assets/images/posts/code.jpg"
+description: ""
+---
+
+## Randomize Your Slack Avatar with Azure Functions: Part 2
+
+Building off [the post last week]({% post_url 2021-07-13-randomize-slack %}), we have expanded the proof of concept to include multiple users and multiple tenants. This example will show how you can utilize Azure Functions to build an interactive Slack App relying on sending and receiving real-time updates to and from Slack.
+
+In this walk-through, we will talk more about how the function app uses Azure resources to operate and interact with Slack. We will not be walking through setting it up. Instructions on how to setup the infrastructure and Slack App are included in the same code repository under the `multi-user-poc` [branch](https://github.com/ocelotconsulting/randomize-avatar/tree/multi-user-poc).
+
+```bash
+git clone https://github.com/ocelotconsulting/randomize-avatar.git
+cd randomize-avatar
+git checkout multi-user-poc
+```
+
+## Slack Interaction Requirements
+
+To make this work, Slack has certain requirements for interactivity:
+
+1. To update the Home Tab for an application, you must send the view definition to [views.publish](https://api.slack.com/methods/views.publish) on a per-user basis
+2. To receive interactive data back, you must supply a `request_url` in the manifest. That URL must:
+   * Be HTTPS with a valid certificate
+   * Not require authentication
+   * Accept `application/x-www-form-urlencoded` POST data which is then deserialized into a JSON object
+   * Must respond with HTTP 200 OK within 3 seconds, otherwise users receive an error message
+
+For our purposes, we update the applications's Home Tab for each user when the avatar is updated (updated timestamp), on initial registration (initial view definition), or on a setting change (updated timestamp).
+
+In our use case, we are building the functions off C#. Unfortunately there is no official [Slack SDK](https://api.slack.com/start/building) built for C# .NET. There is a [community sponsored API](https://github.com/Inumedia/SlackAPI), however it was faster to manually develop the request structure and handle the formatting and JSON manually.
+
+## Our Functions
+
+### UpdateAvatar
+
+This function originated from the `single-user-poc` code base but was reworked to support looking up user information via an [Azure Table](https://docs.microsoft.com/en-us/azure/storage/tables/table-storage-overview) and operate on multiple users at one time. We selected Azure Table storage because it offered a NoSQL option to store user data and the contents are encrypted as rest and in transit. This user data is sensitive enough that it should be protected.
```

Comment:
`encrypted as rest` --> `encrypted at rest`

Reply:
Fixed in 07c2376913b1ca3b83afe806632734ebbeeca1a4